man/sql2csv.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "SQL2CSV" "1" "Aug 16, 2024" "2.2.0" "csvkit"
.SH NAME
sql2csv \- sql2csv Documentation
.SH DESCRIPTION
.sp
Executes arbitrary commands against a SQL database and outputs the results as a CSV:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: sql2csv [\-h] [\-v] [\-l] [\-V] [\-\-db CONNECTION_STRING] [\-\-query QUERY]
               [\-e ENCODING] [\-H]
               [FILE]

Execute a SQL query on a database and output the result to a CSV file.

positional arguments:
  FILE                  The file to use as the SQL query. If FILE and \-\-query
                        are omitted, the query is piped data via STDIN.

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-\-db CONNECTION_STRING
                        A SQLAlchemy connection string to connect to a
                        database.
  \-\-engine\-option ENGINE_OPTION ENGINE_OPTION
                        A keyword argument to SQLAlchemy\(aqs create_engine(), as
                        a space\-separated pair. This option can be specified
                        multiple times. For example: thick_mode True
  \-\-execution\-option EXECUTION_OPTION EXECUTION_OPTION
                        A keyword argument to SQLAlchemy\(aqs
                        execution_options(), as a space\-separated pair. This
                        option can be specified multiple times. For example:
                        stream_results True
  \-\-query QUERY         The SQL query to execute. Overrides FILE and STDIN.
  \-e ENCODING, \-\-encoding ENCODING
                        Specify the encoding of the input query file.
  \-H, \-\-no\-header\-row   Do not output column names.
.ft P
.fi
.UNINDENT
.UNINDENT
.SH EXAMPLES
.sp
Load sample data into a table using \fI\%csvsql\fP and then query it using \fIsql2csv\fP:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvsql \-\-db \(dqsqlite:///dummy.db\(dq \-\-tables \(dqtest\(dq \-\-insert examples/dummy.csv
sql2csv \-\-db \(dqsqlite:///dummy.db\(dq \-\-query \(dqselect * from test\(dq
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Load data about financial aid recipients into PostgreSQL. Then find the three states that received the most, while also filtering out empty rows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
createdb recipients
csvsql \-\-db \(dqpostgresql:///recipients\(dq \-\-tables \(dqfy09\(dq \-\-insert examples/realdata/FY09_EDU_Recipients_by_State.csv
sql2csv \-\-db \(dqpostgresql:///recipients\(dq \-\-query \(dqselect * from fy09 where \e\(dqState Name\e\(dq != \(aq\(aq order by fy09.\e\(dqTOTAL\e\(dq limit 3\(dq
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can even use it as a simple SQL calculator (in this example an in\-memory SQLite database is used as the default):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
sql2csv \-\-query \(dqselect 300 * 47 % 14 * 27 + 7000\(dq
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The connection string \fI\%accepts parameters\fP\&. For example, to set the encoding of a MySQL database:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
sql2csv \-\-db \(aqmysql://user:pass@host/database?charset=utf8\(aq \-\-query \(dqSELECT myfield FROM mytable\(dq
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
Christopher Groskopf and contributors
.SH COPYRIGHT
2016, Christopher Groskopf and James McKinney
.\" Generated by docutils manpage writer.
.


CHANGELOG
0.9.2
-----

* Fix version of openpyxl.

0.9.1
-----

* Add Antonio Lima to AUTHORS.
* Add support for ndjson. (#329)
* Add missing docs for csvcut -C. (#227)
* Reorganize docs so TOC works better. (#339)
* Render docs locally with RTD theme.
* Fix header in "tricks" docs.
* Add install instructions to tutorial. (#331)
* Add killer examples to doc index. (#328)
* Reorganize doc index
* Fix broken csvkit module documentation. (#327)
* Fix version of openpyxl to work around encoding issue. (#391, #288)

0.9.0
-----

* Write missing sections of the tutorial. (#32)
* Remove -q arg from sql2csv (conflicts with common flag).
* Fix csvjoin in case where left dataset rows without all columns.
* Rewrote tutorial based on LESO data. (#324)
* Don't error in csvjson if lat/lon columns are null. (#326)
* Maintain field order in output of csvjson.
* Add unit test for json in2csv. (#77)
* Maintain key order when converting JSON into CSV. (#325.)
* Upgrade python-dateutil to version 2.2 (#304)
* Fix sorting of columns with null values. (#302)
* Added release documentation.
* Fill out short rows with null values. (#313)
* Fix unicode output for csvlook and csvstat. (#315)
* Add documentation for --zero. (#323)
* Fix Integrity error when inserting zero rows in database with csvsql. (#299)
* Add Michael Mior to AUTHORS. (#305)
* Add --count option to CSVStat.
* Implement csvformat.
* Fix bug causing CSVKitDictWriter to output 'utf-8' for blank fields.

0.8.0
-----

* Add pnaimoli to AUTHORS.
* Fix column specification in csvstat. (#236)
* Added "Tips and Tricks" documentation. (#297, #298)
* Add Espartaco Palma to AUTHORS.
* Remove unnecessary enumerate calls. (#292)
* Deprecated DBF support for Python 3+.
* Add support for Python 3.3 and 3.4 (#239)

0.7.3
-----

* Fix date handling with openpyxl > 2.0 (#285)
* Add Kristina Durivage to AUTHORS. (#243)
* Added Richard Low to AUTHORS.
* Support SQL queries "directly" on CSV files. (#276)
* Add Tasneem Raja to AUTHORS.
* Fix off-by-one error in open ended column ranges. (#238)
* Add Matt Pettis to AUTHORS.
* Add line numbers flag to csvlook (#244)
* Only install argparse for Python < 2.7. (#224)
* Add Diego Rabatone Oliveira to AUTHORS.
* Add Ryan Murphy to AUTHORS.
* Fix DBF dependency. (#270)

0.7.2
-----

* Fix CHANGELOG for release.

0.7.1
-----

* Fix homepage url in setup.py.

0.7.0
-----

* Fix XLSX datetime normalization bug. (#223)
* Add raistlin7447 to AUTHORS.
* Merged sql2csv utility (#259).
* Add Jeroen Janssens to AUTHORS.
* Validate csvsql DB connections before parsing CSVs. (#257)
* Clarify install process for Ubuntu. (#249)
* Clarify docs for --escapechar. (#242)
* Make ``import csvkit`` API compatible with ``import csv``.
* Update Travis CI link. (#258)
* Add Sébastien Fievet to AUTHORS.
* Use case-sensitive name for SQLAlchemy (#237)
* Add Travis Swicegood to AUTHORS.

0.6.1
-----

* Add Chris Rosenthal to AUTHORS.
* Fix multi-file input to csvsql. (#193)
* Passing --snifflimit=0 to disable dialect sniffing. (#190)
* Add aarcro to the AUTHORS file.
* Improve performance of csvgrep. (#204)
* Add Matt Dudys to AUTHORS.
* Add support for --skipinitialspace. (#201)
* Add Joakim Lundborg to AUTHORS.
* Add --no-inference option to in2csv and csvsql. (#206)
* Add Federico Scrinzi to AUTHORS file.
* Add --no-header-row to all tools. (#189)
* Fix csvstack blowing up on empty files. (#209)
* Add Chris Rosenthal to AUTHORS file.
* Add --db-schema option to csvsql. (#216)
* Add Shane StClair to AUTHORS file.
* Add --no-inference support to csvsort. (#222)

0.5.0
-----

* Implement geojson support in csvjson. (#159)
* Optimize writing of eight bit codecs. (#175)
* Created csvpy. (#44)
* Support --not-columns for excluding columns. (#137)
* Add Jan Schulz to AUTHORS file.
* Add Windows scripts. (#111, #176)
* csvjoin, csvsql and csvstack will no longer hold open all files. (#178)
* Added Noah Hoffman to AUTHORS.
* Make csvlook output compatible with emacs table markup. (#174)

0.4.4
-----

* Add Derek Wilson to AUTHORS.
* Add Kevin Schaul to AUTHORS.
* Add DBF support to in2csv. (#11, #160)
* Support --zero option for zero-based column indexing. (#144)
* Support mixing nulls and blanks in string columns.
* Add --blanks option to csvsql. (#149)
* Add multi-file (glob) support to csvsql. (#146)
* Add Gregory Temchenko to AUTHORS.
* Add --no-create option to csvsql. (#148)
* Add Anton Ian Sipos to AUTHORS.
* Fix broken pipe errors. (#150)

0.4.3
-----

* Begin CHANGELOG (a bit late, I'll admit).


csvkit/utilities/sql2csv.py
#!/usr/bin/env python
import agate
from sqlalchemy import create_engine

from csvkit.cli import CSVKitUtility, parse_list


class SQL2CSV(CSVKitUtility):
    description = 'Execute a SQL query on a database and output the result to a CSV file.'
    # Overrides all flags except --linenumbers, --verbose, --version.
    override_flags = ['f', 'b', 'd', 'e', 'H', 'I', 'K', 'L', 'p', 'q', 'S', 't', 'u', 'z', 'zero','add-bom']

    def add_arguments(self):
        self.argparser.add_argument(
            '--db', dest='connection_string', default='sqlite://',
            help='A SQLAlchemy connection string to connect to a database.')
        self.argparser.add_argument(
            '--engine-option', dest='engine_option', nargs=2, action='append', default=[],
            help="A keyword argument to SQLAlchemy's create_engine(), as a space-separated pair. "
                 "This option can be specified multiple times. For example: thick_mode True")
        self.argparser.add_argument(
            '--execution-option', dest='execution_option', nargs=2, action='append',
            # https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection.execution_options.params.no_parameters
            # https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection.execution_options.params.stream_results
            # https://docs.sqlalchemy.org/en/20/core/connections.html#using-server-side-cursors-a-k-a-stream-results
            default=[['no_parameters', True], ['stream_results', True]],
            help="A keyword argument to SQLAlchemy's execution_options(), as a space-separated pair. "
                 "This option can be specified multiple times. For example: stream_results True")
        self.argparser.add_argument(
            metavar='FILE', nargs='?', dest='input_path',
            help='The file to use as SQL query. If FILE and --query are omitted, the query is piped data via STDIN.')
        self.argparser.add_argument(
            '--query',
            help="The SQL query to execute. Overrides FILE and STDIN.")
        self.argparser.add_argument(
            '-e', '--encoding', dest='encoding', default='utf-8',
            help='Specify the encoding of the input query file.')
        self.argparser.add_argument(
            '-H', '--no-header-row', dest='no_header_row', action='store_true',
            help='Do not output column names.')

        self.argparser.set_defaults(
            delimiter=None,
            doublequote=None,
            escapechar=None,
            encoding='utf-8',
            field_size_limit=None,
            quotechar=None,
            quoting=None,
            skipinitialspace=None,
            tabs=None,
        )

    def main(self):
        if self.additional_input_expected() and not self.args.query:
            self.argparser.error('You must provide an input file or piped data.')

        try:
            engine = create_engine(self.args.connection_string, **parse_list(self.args.engine_option))
        except ImportError as e:
            raise ImportError(
                "You don't appear to have the necessary database backend installed for connection string you're "
                "trying to use. Available backends include:\n\nPostgreSQL:\tpip install psycopg2\nMySQL:\t\tpip "
                "install mysql-connector-python OR pip install mysqlclient\n\nFor details on connection strings "
                "and other backends, please see the SQLAlchemy documentation on dialects at:\n\n"
                "https://www.sqlalchemy.org/docs/dialects/"
            ) from e

        connection = engine.connect()

        if self.args.query:
            query = self.args.query.strip()
        else:
            query = ""

            self.input_file = self._open_input_file(self.args.input_path)

            for line in self.input_file:
                query += line

            self.input_file.close()

        rows = connection.execution_options(**parse_list(self.args.execution_option)).exec_driver_sql(query)
        output = agate.csv.writer(self.output_file, **self.writer_kwargs)

        if rows.returns_rows:
            if not self.args.no_header_row:
                output.writerow(rows._metadata.keys)

            for row in rows:
                output.writerow(row)

        connection.close()
        engine.dispose()


def launch_new_instance():
    utility = SQL2CSV()
    utility.run()


if __name__ == '__main__':
    launch_new_instance()
